8255119: Monitor::wait takes signed integer as timeout

dholmes-ora opened the pull request:
Please review this simple cleanup of the JVM `Monitor`/`PlatformMonitor` code so that timeouts are expressed as unsigned rather than signed values i.e. `int64_t`/`jlong` -> `uint64_t`.

Testing:
 - tiers 1-3
 - all builds in Oracle CI in tiers 1-5, plus shenandoah GC on LInux x64

Thanks

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8255119](https://bugs.openjdk.org/browse/JDK-8255119): Monitor::wait takes signed integer as timeout


### Reviewers
 * [Johan Sjölen](https://openjdk.org/census#jsjolen) (@jdksjolen - Committer) ⚠️ Review applies to [674892df](https://git.openjdk.org/jdk/pull/11898/files/674892df036866f991069bb63ee99a00d77e590e)
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**) ⚠️ Review applies to [674892df](https://git.openjdk.org/jdk/pull/11898/files/674892df036866f991069bb63ee99a00d77e590e)
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to [674892df](https://git.openjdk.org/jdk/pull/11898/files/674892df036866f991069bb63ee99a00d77e590e)
 * [Gerard Ziemski](https://openjdk.org/census#gziemski) (@gerard-ziemski - Committer) ⚠️ Review applies to [674892df](https://git.openjdk.org/jdk/pull/11898/files/674892df036866f991069bb63ee99a00d77e590e)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11898/head:pull/11898` \
`$ git checkout pull/11898`

Update a local copy of the PR: \
`$ git checkout pull/11898` \
`$ git pull https://git.openjdk.org/jdk pull/11898/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11898`

View PR using the GUI difftool: \
`$ git pr show -t 11898`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11898.diff">https://git.openjdk.org/jdk/pull/11898.diff</a>

</details>

jdksjolen reviewed (approved):
Looks good to me, thanks.

coleenp commented on src/hotspot/os/windows/os_windows.cpp:
```diff
@@ -5358,9 +5358,11 @@ void Parker::unpark() {
-  assert(millis >= 0, "negative timeout");
+int PlatformMonitor::wait(uint64_t millis) {
   int ret = OS_TIMEOUT;
+  if (millis > UINT_MAX) {
```
Can you add a one line comment saying why this is?  is it a platform limitation? or just would be mostly infinite if bigger than UINT_MAX?

dholmes-ora replied:
The platform API actually takes a DWORD aka uint32_t. I will add a comment.

gerard-ziemski commented:
LGTM

Looks like there are other APIs where we use signed values for the time, which is illegal - do we have an issue tracking that cleanup?

```
int PlatformEvent::park(jlong millis) {
  // Transitions for _event:
  //   -1 => -1 : illegal
  //    1 =>  0 : pass - return immediately
  //    0 => -1 : block; then set _event to 0 before returning
```

dholmes-ora commented:
> Looks like there are other APIs where we use signed values for the time, which is illegal - do we have an issue tracking that cleanup?

No, as per earlier comments in this bug report that code is fed by signed values that come from Java. We could impose a signed -> unsigned conversion at the VM boundary, but then we also have to redo all the  conversion routines to pack and unpack that time structs that we use. I don't see any point in doing all that.

dholmes-ora commented:
Thanks for the reviews @jdksjolen , @tstuefe , @coleenp  and @gerard-ziemski . I will add the requested comment and integrate.

dholmes-ora commented:
I realized the windows code still had a pre-existing truncation bug due to the 64-bit -> 32-bit conversion. In practice it is not an issue as we never use such large timeout values in the VM. Re-running Windows builds/tests.

dholmes-ora commented:
> I realized the windows code still had a pre-existing truncation bug due to the 64-bit -> 32-bit conversion. In practice it is not an issue as we never use such large timeout values in the VM. Re-running Windows builds/tests.

Strike that. I fixed the truncation bug with the UINT_MAX check. All that was missing was an explicit cast to DWORD - which for some reason the compiler never complained about, but I added for clarity.

dholmes-ora commented:
/integrate
